website/docs/language/expressions/custom-validation-rules.mdx

```diff
@@ -0,0 +1,316 @@
+---
+page_title: Custom Validation Rules - Configuration Language
+description: >-
+  Validate requirements for variables, outputs, and within lifecycle blocks so Terraform can produce better error messages in context.
+---
+
+# Custom Validation Rules
+
+You can create validation checks with custom error messages for several types of objects in a configuration. Custom validations are optional, but they can help capture assumptions that might be only implied, helping future maintainers understand the configuration design and intent. They also return useful information about errors earlier and in context, helping consumers more easily diagnose issues in their configurations.
+
+You can create custom validations with the following types of expressions.
+- Add [`validation` blocks](input-variable-validation) inside input `variable` blocks.
+- Add [`precondition` and `postcondition`](#preconditions-and-postconditions) blocks inside `lifecycle` blocks and `precondition` blocks inside `output blocks`.
```

Comment:
I worry that the reference to `lifecycle` blocks here (and in the page description) rather buries the lede that these validations can be applied to resources and data sources.

Reply:
This is a good call! I'm going to update to emphasize the object types a bit more.